src/lib/extension-load-queue.ts

```diff
@@ -1,63 +1,149 @@
+ *   store: myStore,
+ *   createController: () => new MyController(),
+ *   queueKey: MY_QUEUE_SYMBOL,
+ * });
+ *
+ * const controller = useController(myStore);
+ */
+
 import {useEffect, useState} from 'react';
 
+import {isBrowser} from './browser';
+
 export type ControllerLoadedCallback<T> = (controller: T) => void;
 
-export const QUEUE_SYMBOL = Symbol.for('queue');
+export const QUEUES_SYMBOL = Symbol.for('extension-load-queues');
+export const SINGLE_QUEUE_SYMBOL = Symbol.for('single-queue');
 
 export type ScriptStore<T> = ControllerLoadedCallback<T>[];
 
+/**
+ * Interface for creating a load queue.
+ */
 export interface CreateLoadQueueArgs<T> {
+    // The store where callbacks are saved
     store: ScriptStore<T>;
+
+    // Function that creates a controller
     createController: () => T;
+
+    // Flag to check if the queue is already created
     isQueueCreated?: boolean;
+
+    // Callback to handle queue creation
     onQueueCreated?: (created: boolean) => void;
-}
 
-// TODO: this function is a very frequently used utility,
-//  so we should place it in a separate top-level file and document it.
-export const isBrowser = () => typeof window !== 'undefined' && typeof document !== 'undefined';
+    // Symbol key for identifying the queue
+    queueKey?: symbol;
+}
 
-// TODO: window casts are weird — fix encounter any issues with declaration merging
-export const getScriptStore = <T>(key: symbol): ScriptStore<T> => {
+/**
+ * Retrieves or initializes a script store for the given symbol key.
+ *
+ * @param {symbol} storeKey - The key associated with the script store.
+ * @returns {ScriptStore<T>} The script store associated with the given key.
+ */
+export const getScriptStore = <T>(storeKey: symbol): ScriptStore<T> => {
     if (isBrowser()) {
-        (window as Window)[key] = (window as Window)[key] || [];
-        return (window as Window)[key];
+        // Initialize the store if it does not exist or check that it is an array
+        if (!window[storeKey]) {
+            window[storeKey] = [];
+        } else if (!Array.isArray(window[storeKey])) {
+            throw new Error(`Expected window[${String(storeKey)}] to be an array`);
+        }
+
+        return window[storeKey] as ScriptStore<T>;
     } else {
         throw new Error('This functionality should be employed on the client-side.');
     }
 };
 
-export const getQueueStore = () => {
+/**
+ * Ensures that the queues are initialized as an object in the global window.
+ *
+ * @returns {void}
+ */
+const ensureQueuesSymbolInitialized = (): void => {
     if (isBrowser()) {
-        (window as Window)[QUEUE_SYMBOL] = (window as Window)[QUEUE_SYMBOL] || false;
-        return (window as Window)[QUEUE_SYMBOL];
+        if (typeof window[QUEUES_SYMBOL] !== 'object' || window[QUEUES_SYMBOL] === null) {
+            window[QUEUES_SYMBOL] = {};
+        }
+    } else {
+        throw new Error('This functionality should be employed on the client-side.');
```

Comment:
I don't think this error message is self-descriptive enough. The message should mention that it comes from a load queue. Maybe something along the lines of `Cannot initialize QueueStore in a non-browser environment`?